runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java

```diff
@@ -139,6 +144,19 @@ public TimerInternals timerInternalsForKey(String key) {
           }
         });
 
+    final ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor(
```

Comment:
Including this as a local variable in the `createEvaluator` call was a mistake (which I can see we've tried to reduce the impact of by making sure all the threads are daemons, which means that they won't block shutdown; still, it's creating a long-lived object per execution rather than per-enclosing-long-lived-object)

Instead, move it to be a factory-wide cached threadpool executor, which can be shutdown in `cleanup` without having to alter the `ProcessFn` implementation, or take on `BEAM-723` and provide a shared `ExecutorService` in `ExecutionContext` (which can be shut down as part of the overall termination of the runner)

Also, we should be able to cache `processFn` instances for the life of the pipeline, I think. @jkff any reason that wouldn't be possible? It looks like we'll set all of the information scoped per-bundle as appropriate and the remaining (unchanged) elements are immutable (or can be cached at the same granularity the `ProcessFn` would be cached at)

Reply:
processFn has to be at least thread-local because it's not thread-safe; but other than that yeah, if there's a way to make it live throughout the life of the pipeline, that'd be fine with me. The current implementation is just for simplicity. I think I tried to make processFn more global, and hit some difficulty, but I no longer remember what.

Reply:
Updating for ses

For processFn it is a bit more complex - and surely outside the scope of this PR - since it requires the CommittedBundle so it needs to introduce a real lifecycle management there is not yet (init etc).